### PR TITLE
fix: Make peers sync secondary index

### LINKED ISSUE #2379: Secondary indexes does not appear to work with P2P
For example, the final action below fails as the query returns no results.  Deleting the `CreateIndex` test action causes the test to pass.

The order in which the fields are declared in the index does not affect the results.  Using a single-field index still results in failure, suggesting this affects 0.9.0 and all indexes on P2P synced data.

Test is not commited to develop branch.

```
func TestP2PCreateWithP2PCollection2(t *testing.T) {
	test := testUtils.TestCase{
		Actions: []any{
			testUtils.RandomNetworkingConfig(),
			testUtils.RandomNetworkingConfig(),
			testUtils.SchemaUpdate{
				Schema: `
					type Users {
						Name: String
						Age: Int
					}
				`,
			},
			testUtils.CreateIndex{
				CollectionID: 0,
				IndexName:    "name_age_index",
				Fields:       []testUtils.IndexedField{{Name: "Name"}, {Name: "Age"}},
			},
			testUtils.ConnectPeers{
				SourceNodeID: 1,
				TargetNodeID: 0,
			},
			testUtils.SubscribeToCollection{
				NodeID:        1,
				CollectionIDs: []int{0},
			},
			testUtils.CreateDoc{
				NodeID: immutable.Some(0),
				Doc: `{
					"Name": "Fred",
					"Age": 31
				}`,
			},
			testUtils.WaitForSync{},
			testUtils.Request{
				NodeID: immutable.Some(1),
				Request: `query {
					Users (filter: {Age: {_eq: 31}, Name: {_eq: "Fred"}}){
						Age
					}
				}`,
				Results: []map[string]any{
					{
						"Age": int64(31),
					},
				},
			},
		},
	}

	testUtils.ExecuteTestCase(t, test)
}
```

### PR DESCRIPTION
**Pull request by islamaliev**
## Relevant issue(s)

Resolves #2379

## Description

Make peers updated update local storage when an update happens on docs.

### CODE REVIEW

File: client/db.go [hunk 1]
```diff
@@ -215,6 +215,15 @@ type Store interface {
 	// GetAllIndexes returns all the indexes that currently exist within this [Store].
 	GetAllIndexes(context.Context) (map[CollectionName][]IndexDescription, error)
 
+	// CreateDocIndex creates an index for the given document.
+	CreateDocIndex(context.Context, Collection, *Document) error
```
Comment (AndrewSisley) [inline client/db.go:219]: suggestion: Are you sure you want 3 public functions for this?  Is there not only ever one valid index state for a document at any given time, and would this not be more safely handled by a `RefreshIndex(doc)`-like function, that decides for the caller what underlying action(s) should be taken?

Comment (islamaliev) [reply client/db.go:219]: it's not possible. Indexes operate on values, on ids. To update an indexed document you need to provide a doc with the old value to delete from index and a doc with a new value to add.
Proposed `RefreshIndex` receives only one doc.

Comment (AndrewSisley) [reply client/db.go:219]: Ahhh, true - that's a bit annoying.  I wonder if we can get around that somehow by using something clever like tracking the txn or using events, but that can be looked at later, not in the last couple of days before release :)

Thanks for the explanation :)

Comment (AndrewSisley) [inline client/db.go:219]: suggestion: It feels like hosting these functions on `client.Collection` would be nicer, given that they have to take a `Collection` as a param anyway.

File: db/collection_index.go [hunk 2]
```diff
@@ -107,6 +107,26 @@ func (db *db) fetchCollectionIndexDescriptions(
 	return indexDescriptions, nil
 }
 
+func (db *db) getCollectionIndexes(
+	ctx context.Context,
+	txn datastore.Txn,
+	col client.Collection,
+) ([]CollectionIndex, error) {
+	indexDescriptions, err := db.fetchCollectionIndexDescriptions(ctx, txn, col.ID())
```
Comment (AndrewSisley) [inline db/collection_index.go:115]: todo: Are you sure this is necessary?  The IndexDescriptions should already be on the `col.Description()` that you are passing into this function.

File: client/db.go (hunk 1 above)
Comment (islamaliev) [reply client/db.go:219]: that was our initial approach (Fred also took part in this design), but with all these http and cli wrappers it turned out to be more complicated. That's why wee took this route

File: db/txn_db.go [hunk 3]
```diff
@@ -174,14 +174,154 @@ func (db *implicitTxnDB) GetAllIndexes(
...
+		return err
+	}
+	defer txn.Discard(ctx)
+
+	indexes, err := db.getCollectionIndexes(ctx, txn, col)
+	if err != nil {
+		return err
+	}
+
+	for _, index := range indexes {
```
Comment (AndrewSisley) [inline db/txn_db.go:204]: todo: Please don't include any non transaction related logic in this file - instead move this block into a `db.createDocIndex` func (along with the others).

This allows us to keep the transaction boiler-plate stuff separate from the more complex business logic.

File: http/client.go [hunk 4]
```diff
@@ -434,3 +434,28 @@ func (c *Client) Events() events.Events {
 func (c *Client) MaxTxnRetries() int {
 	panic("client side database")
 }
+
+func (c *Client) CreateDocIndex(
+	ctx context.Context,
+	col client.Collection,
+	doc *client.Document,
+) error {
+	panic("client side database")
```
Comment (AndrewSisley) [inline http/client.go:443]: todo: Please implement these :)

File: tests/clients/cli/wrapper.go [hunk 5]
```diff
@@ -522,3 +522,28 @@ func (w *Wrapper) WaitForPushLogByPeerEvent(id peer.ID) error {
 func (w *Wrapper) WaitForPushLogFromPeerEvent(id peer.ID) error {
 	return w.node.WaitForPushLogFromPeerEvent(id)
 }
+
+func (w *Wrapper) CreateDocIndex(
+	ctx context.Context,
+	col client.Collection,
+	doc *client.Document,
+) error {
+	panic("client side database")
```
Comment (AndrewSisley) [inline tests/clients/cli/wrapper.go:531]: todo: Please implement these :)

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
Looks good, but I think there are few bits that need sorting out before merge

File: client/db.go (hunk 1 above)
Comment (AndrewSisley) [reply client/db.go:219]: Why was that more complicated?  The bulk of the logic for handling collection stuff already exists in http/cli

File: db/txn_db.go (hunk 3 above)
Comment (islamaliev) [reply db/txn_db.go:204]: done

File: tests/clients/cli/wrapper.go (hunk 5 above)
Comment (islamaliev) [reply tests/clients/cli/wrapper.go:531]: this is not supposed to be called from outside

File: http/client.go (hunk 4 above)
Comment (AndrewSisley) [reply http/client.go:443]: We have plenty of http/cli funcs that receive docs as params

File: tests/clients/cli/wrapper.go (hunk 5 above)
Comment (AndrewSisley) [reply tests/clients/cli/wrapper.go:531]: Then it shouldn't be in the `client` package

File: db/collection_index.go (hunk 2 above)
Comment (islamaliev) [reply db/collection_index.go:115]: you are right. Adjusted. Nice catch!

File: client/db.go (hunk 1 above)
Comment (islamaliev) [reply client/db.go]: moved

File: tests/clients/cli/wrapper.go (hunk 5 above)
Comment (islamaliev) [reply tests/clients/cli/wrapper.go]: Added error as discussed in syncup

File: http/client.go (hunk 4 above)
Comment (islamaliev) [reply http/client.go]: Added error as discussed in syncup

File: db/collection_index.go (hunk 2 above)
Comment (AndrewSisley) [reply db/collection_index.go]: nice! thanks :)

File: client/collection.go [hunk 6]
```diff
@@ -169,6 +169,21 @@ type Collection interface {
...
 	// GetIndexes returns all the indexes that exist on the collection.
 	GetIndexes(ctx context.Context) ([]IndexDescription, error)
+
+	// CreateDocIndex creates an index for the given document.
+	// WARNING: This method is only for internal use and is not supposed to be called by the client
+	// as it might compromise the integrity of the database. This method will be removed in the future
+	CreateDocIndex(context.Context, *Document) error
+
+	// UpdateDocIndex updates the index for the given document.
+	// WARNING: This method is only for internal use and is not supposed to be called by the client
```
Comment (AndrewSisley) [inline client/collection.go:179]: praise: These are suitably scary - thanks :grin:

Comment (fredcarle) [reply client/collection.go:179]: I second that 👍

File: db/collection_index.go [hunk 7]
```diff
@@ -107,6 +107,55 @@ func (db *db) fetchCollectionIndexDescriptions(
...
+	if err != nil {
+		return err
+	}
+	defer c.discardImplicitTxn(ctx, txn)
+
+	err = c.deleteIndexedDoc(ctx, txn, oldDoc)
+	if err != nil {
+		return err
+	}
+	err = c.indexNewDoc(ctx, txn, newDoc)
```
Comment (AndrewSisley) [inline db/collection_index.go:136]: nitpick: I do see `delete+new` as still being business logic, but only at a nitpick level - so feel free to merge as is :)

**Review by fredcarle — APPROVED** [review]
LGTM!

File: net/server.go [hunk 8]
```diff
@@ -338,6 +327,64 @@ func (s *server) PushLog(ctx context.Context, req *pb.PushLogRequest) (*pb.PushL
...
+		},
+	)
+	if err != nil {
+		return nil, errors.Wrap(fmt.Sprintf("Failed to get collection from schemaRoot %s", schemaRoot), err)
+	}
+	if len(cols) == 0 {
+		return nil, client.NewErrCollectionNotFoundForSchema(schemaRoot)
+	}
+	var col client.Collection
+	for _, c := range cols {
```
Comment (AndrewSisley) [inline net/server.go:348]: praise: Thanks for moving this somewhere out of the way :)

**Review by AndrewSisley — APPROVED** [review]
LGTM - thanks Islam :)

File: db/collection_index.go (hunk 7 above)
Comment (fredcarle) [reply db/collection_index.go:136]: Can you expand on what is your nitpick concern here Andy?

Comment (AndrewSisley) [reply db/collection_index.go:136]: I have none and thought I was in the txn_db.go file :)

Comment (AndrewSisley) [reply db/collection_index.go:136]: Ah I see I how I got confused - they have moved to collection, and I new that they had moved, but my brain still thought of the implementation as being in the txn_db file when I saw the txn logic.  I wonder if we can do a similar job to collection that we did to client.Store and split out the txn boilerplate to another file sometime

**Pull request merged**